File: ref/csp/templates/python/3-problems/1-searching/m-sum.py
#!/usr/bin/env python3

import sys

input = sys.stdin.readline

# n -> sum of nums
# m -> count of nums
n, m = map(int, input().split())

path = []
res = []


def dfs(pos, rsrc, prev):
    if pos > m:
        if rsrc == 0:
            res.append(path[:])
        return

    for x in range(prev + 1, n + 1):
        if rsrc - x < 0:
            break

        k = m - pos
        min_sum = k * (2 * x + k + 1) // 2

        if rsrc - x - min_sum < 0:
            break

        path.append(x)
        dfs(pos + 1, rsrc - x, x)
        path.pop()


dfs(1, n, 0)
print(res)

File: ref/csp/templates/cpp/3-searching/m-sum.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<vector<int>> res;
    vector<int> path;

    auto f = [&](auto &&self, int pos, int rsrc, int prev) -> void {
        if (pos > m) {
            if (rsrc == 0) {
                res.push_back(path);
            }
            return;
        }

        for (int x = prev + 1; rsrc - x >= 0; x++) {
            int k = m - pos;
            int min_sum = k * (2 * x + k + 1) / 2;

            if (rsrc - x - min_sum < 0) {
                break;
            }

            path.push_back(x);
            self(self, pos + 1, rsrc - x, x);
            path.pop_back();
        }
    };

    f(f, 1, n, 0);

    for (auto p : res) {
        for (auto x : p) {
            cout << x << " ";
        }
        cout << "\n";
    }
}
